### DISCUSSION TITLE
How to push tags ?

### DISCUSSION THREAD

**Discussion by zmitya** [jesseduffield/lazygit#2480] [General]
Gents, sorry for the newbie question, but could not find the shortcut which equvivalent to

git push origin --tags

Command...
Thanks,
Mitya

**Comment by mark2185** [comment]
Go to the `Local Branches` panel and move to the `Tags` tab (pressing `]`), there you can open the `help` menu with `?`

It states:
`P       push tag`

**Reply by zmitya** [reply]
Wow, thank you very much, really appreciate it...
Please mark tha above as an answer, I can't do it :(

**Reply by ShortCipher** [reply]
Yes, thank you @mark2185!  I couldn't figure this out either. 😂 
I kept going back to the shell to write `$ git push --tags` 🤦🏼‍♂️

**Reply by mark2185** [reply]
Great!

I hope you were at least using the `:` mapping so the command was in your suggestions :)

**Reply by mmolhoek** [reply]
Does the P form the tags tab also push the commit that I created the tag on?

**Reply by mmolhoek** [reply]
it does not. so you need to push commits, then push tags

**Reply by LintaoAmons** [reply]
Could we have another shortcut in the commits panel so it's much easier to do it?

Currently, workflow is too cumbersome
1. I commit
2. Go to the commits panel and use `T` to tag the latest commit
3. use `P` to push the commit
4. navigate to the `tag` panel
5. then find the tag and use `P` to push to remote

I think maybe it would be better to just.

1. I commit
2. Go to the commits panel and use `T` to tag the latest commit
3. use `P` to push the commit and the tags related to this commit

**Reply by stefanhaller** [reply]
I haven't tried this myself, but from reading the git documentation it seems that you get this behaviour automatically by doing `git config --global push.followTags true`. Have you tried that?

**Reply by LintaoAmons** [reply]
I tried, but it seems like not working.

**Reply by stefanhaller** [reply]
According to the documentation, it works for annotated tags; it doesn't talk about lightweight tags. Is that what you tried?

**Reply by LintaoAmons** [reply]
Yeah, but I think that's the kind of tag been created by the `T` shortcut

**Reply by stk-ableton** [reply]
The `T` shortcut creates a lightweight tag if you leave the description empty, or an annotated tag if you don't.

**Reply by LintaoAmons** [reply]
Ehh, I'm actually didn't know the difference between the lightweight tag and the annotated tag~

Normally I would just create lightweight tags...

I shall learn it first and see if annotated one could be pushed along with the commits.

Though, I thought maybe it's a good idea to have an option to allow users push lightweight tags when `P` in the commits panel. Sound like a simple workflow~

**Reply by stefanhaller** [reply]
In my opinion, lightweight tags are for local use only. I sometimes use them for short-term use, e.g. I'll create one before I start a complicated rebase, so that I can easily go back to it when the rebase experiments don't work out.

Tags that you publish should always be annotated, because then they contain information about who created the tag and when.

**Reply by LintaoAmons** [reply]
Thanks for your patient explanation,
I didn't notice that before ~

**Reply by doga** [reply]
Must press `]` _twice_ to go to the `Tags` tab when you are in `Local branches`. Pressing once only takes you to the `Remotes` tab.

**Reply by LintaoAmons** [reply]
> Must press `]` _twice_ to go to the `Tags` tab when you are in `Local branches`. Pressing once only takes you to the `Remotes` tab.

Yeah, I still find it inconvenient.
And to add an annotated tag I also need to press tab and type something then tab back then enter. Even though my annotation is meaningless string just to make it an annotated tag

**Reply by doga** [reply]
> to add an annotated tag I also need to press tab and type something then tab back then enter.

Agreed, that's a UX issue. Same thing with commits. This is worth submitting a new issue, maybe someone already did, haven't checked.

**Reply by stefanhaller** [reply]
For tags, I wouldn't be opposed to adding a user config like `git.alwaysMakeAnnotatedTags=true`, which would pass `-a` and an empty message if the user didn't enter one.

What's the issue with commits though? They don't have this distinction between lightweight and annotated, so I'm not sure how they are related. @doga Care to elaborate?

**Reply by doga** [reply]
> What's the issue with commits though?

Just added this issue to the issues list: [UX for committing and tagging is non-obvious #4134](https://github.com/jesseduffield/lazygit/issues/4134)
